susemanager-utils/susemanager-sls/src/_modules/uyuni_users.py

```diff
@@ -0,0 +1,825 @@
+
+    def save_session(self) -> bool:
+        """
+        Save session of the RPC
+
+        :return: boolean, True on success
+        """
+        os.makedirs(os.path.dirname(RPCClient.__instance_path__), exist_ok=True)
+        with open(RPCClient.__instance_path__, 'wb') as fh:
+            try:
+                fh.write(str(self.get_token()).encode())
+                log.debug("Wrote RPC session to %s", RPCClient.__instance_path__)
+                ret = True
+            except OSError as exc:
+                ret = False
+                log.error("Unable to serialise RPC client: %s", exc)
+            except Exception as exc:
+               ret = False
+               log.error("Unhandled error while serialising RPC client object: %s", exc)
+
+        return ret
+
+    @staticmethod
+    def load_session() -> Optional[str]:
+        """
+        Read previously saved RPC session token from the disk.
+        If this is not possible, None is returned.
+
+        :return: string or None
+        """
+        obj: Optional[str] = None
+        try:
+            with open(RPCClient.__instance_path__, 'rb') as fh:
+                try:
+                    data: bytes = fh.read()
+                    if data:
+                        obj = data.decode()
+                except Exception as exc:
+                    log.debug("Unable to load saved RPC session: %s", exc)
+        except FileNotFoundError:
+            log.debug("No previously saved RPC session")
+
+        return obj
+
+    @staticmethod
+    def init(pillar: Optional[Dict[str, Any]] = None):
+        """
+        Create new instance
+
+        :return:
+        """
+        if RPCClient.__instance__ is None:
+            plr: Optional[Dict[str, Any]] = __pillar__ or {}
+            if "xmlrpc" not in (plr or {}).keys():
+                plr = pillar
+
+            if "xmlrpc" in (plr or {}).get("uyuni", {}):
+                rpc_conf = (plr or {})["uyuni"]["xmlrpc"] or {}
+                RPCClient.__instance__ = RPCClient(rpc_conf.get("url", "https://localhost/rpc/api"),
+                                                   rpc_conf.get("user", ""), rpc_conf.get("password", ""))
+                RPCClient.__instance__.token = RPCClient.load_session()
+            else:
+                raise UyuniUsersException("Unable to find Pillar configuration for Uyuni XML-RPC API")
+
+        return RPCClient.__instance__
+
+    def get_token(self, refresh: bool = False) -> Optional[str]:
+        """
+        Authenticate.
+
+        :return:
+        """
+        if self.token is None or refresh:
+            try:
+                self.token = self.conn.auth.login(self._user, self._password)
+                self.save_session()
+            except Exception as exc:
+                log.error("Unable to login to the Uyuni server: %s", exc)
+
+        return self.token
+
+    def __call__(self, method: str, *args, **kwargs) -> Any:
+        self.get_token()
```

Comment:
It seems to me that all callers of this method already call `get_token`, is there a reason I cannot presently see why is the call repeated here?

Reply:
It is double-call, if token expires.